jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -152,13 +193,465 @@ default int capacity()
+            }
+        }
+
+        @Override
+        public boolean canRetain()
+        {
+            return _buffer.canRetain();
+        }
+
+        @Override
+        public boolean isRetained()
+        {
+            return _buffer.isRetained();
+        }
+
+        @Override
+        public void retain()
+        {
+            _buffer.retain();
+        }
+
+        @Override
+        public boolean release()
+        {
+            return _buffer.release();
+        }
+
+        @Override
+        public ByteBuffer getByteBuffer()
+        {
+            return _buffer.getByteBuffer();
+        }
+
+        @Override
+        public RetainableByteBuffer copy()
+        {
+            RetainableByteBuffer buffer = _buffer;
+            buffer.retain();
+            return new AbstractRetainableByteBuffer(buffer.getByteBuffer().slice())
+            {
+                {
+                    acquire();
+                }
+
+                @Override
+                public boolean release()
+                {
+                    if (super.release())
+                    {
+                        buffer.release();
+                        return true;
+                    }
+                    return false;
+                }
+            };
+        }
+
+        @Override
+        public int capacity()
+        {
+            return Math.max(_buffer.capacity(), _maxCapacity);
+        }
+
+        @Override
+        public boolean append(ByteBuffer bytes)
+        {
+            ensureSpace(bytes.remaining());
+            return RetainableByteBuffer.super.append(bytes);
+        }
+
+        @Override
+        public boolean append(RetainableByteBuffer bytes)
+        {
+            ensureSpace(bytes.remaining());
+            return RetainableByteBuffer.super.append(bytes);
+        }
+
+        private void ensureSpace(int spaceNeeded)
+        {
+            int capacity = _buffer.capacity();
+            int space = capacity - _buffer.remaining();
+            if (spaceNeeded <= space || capacity >= _maxCapacity)
+                return;
+
+            int newCapacity = Math.multiplyExact(1 + Math.addExact(capacity, spaceNeeded) / _growBy,  _growBy);
+            if (newCapacity > _maxCapacity)
+            {
+                newCapacity = Math.addExact(capacity, spaceNeeded - space);
+                if (newCapacity > _maxCapacity)
+                    newCapacity = _maxCapacity;
+            }
+
+            RetainableByteBuffer ensured = _pool.acquire(newCapacity, _direct);
+            ensured.append(_buffer);
+            _buffer.release();
+            _buffer = ensured;
+        }
+    }
+
+    /**
+     * An accumulating {@link RetainableByteBuffer} that may internally accumulate multiple other
+     * {@link RetainableByteBuffer}s with zero-copy if the {@link #append(RetainableByteBuffer)} API is used
+     */
+    class Accumulator implements RetainableByteBuffer
```

Comment:
I don't think this is right.

Like `Aggregator` it should _have_ a RBB, not _be_ one.

It breaks Liskov, and it's obvious that it must have its own separate API because it can store more than 2 GiB of buffers, but if you do so, all the methods will throw trying to copy the bytes into a single BB.

Reply:
See Aggregator comments.

Also, it is silly to say that since zero-copy accumulation of buffers may allow the total length to exceed 2GB, then we cannot use that efficient implementation for appenders that will not exceed 2GB.

Yes there is an issue of callers try to get a single BB from the API, but the whole point of hiding the implementation is that we should provide suitable convenience methods that will make explicit calls to `getByteBuffer` an anti-pattern. For example we should call `rbb.writeTo(sink, true, callback)` rather than `sink.write(rbb.getByteBuffer(), true, callback)`

Because we do not provide the correct APIs, we have lots of copies happening now when we could avoid them.